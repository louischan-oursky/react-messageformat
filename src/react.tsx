import * as React from "react";
import { Token } from "@louischan-oursky/messageformat-parser";
import { parse } from "./parse";
import { localeToLanguage } from "./locale";
import { Values, evaluate } from "./eval";

export interface Components {
  [key: string]: React.ReactType;
}

export interface ContextValue {
  locale: string;
  language: string;
  compile: (id: string) => Token[];
  renderToString: (id: string, values?: Values) => string;
}

export interface ProviderProps {
  locale: string;
  messageByID: { [key: string]: string | undefined };
  children?: React.ReactNode;
}

interface ProviderState {
  version: number;
}

export interface MessageOwnProps {
  id: string;
  values?: Values;
  components?: Components;
}

interface MessageContextProps {
  language: string;
  compile: (id: string) => Token[];
}

type MessageProps = MessageOwnProps & MessageContextProps;

const defaultValue: ContextValue = {
  locale: "en",
  language: "en",
  compile: () => [],
  renderToString: () => "",
};
const Context = React.createContext<ContextValue>(defaultValue);
const { Provider, Consumer } = Context;

export { Context };
export { Consumer };

export class LocaleProvider extends React.Component<
  ProviderProps,
  ProviderState
> {
  tokensByID: { [key: string]: Token[] | undefined } = {};
  state = {
    version: 0,
  };

  componentDidUpdate(prevProps: ProviderProps) {
    if (this.props.locale !== prevProps.locale) {
      this.tokensByID = {};
      this.setState((prevState) => {
        return {
          version: prevState.version + 1,
        };
      });
    }
  }

  compile = (id: string) => {
    const cachedTokens = this.tokensByID[id];
    if (cachedTokens != null) {
      return cachedTokens;
    }
    const source = this.props.messageByID[id];
    if (source == null) {
      throw new Error(`expected "${id}" to exist`);
    }
    try {
      const tokens = parse(source);
      this.tokensByID = {
        ...this.tokensByID,
        [id]: tokens,
      };
      return tokens;
    } catch (e) {
      const message = e.message;
      e.message = `"${id}": ${message}`;
      throw e;
    }
  };

  renderToString = (id: string, values?: Values) => {
    try {
      const { locale } = this.props;
      const tokens = this.compile(id);
      const language = localeToLanguage(locale);
      const result = evaluate(tokens, language, values || {}, {});
      const output = [];
      for (const a of result) {
        if (typeof a !== "string") {
          throw new Error(`unexpected non-string value "${id}"`);
        }
        output.push(a);
      }
      return output.join("");
    } catch (e) {
      console.warn(e);
      return "";
    }
  };

  render() {
    const { locale, children } = this.props;
    return (
      <Provider
        value={{
          locale,
          language: localeToLanguage(locale),
          compile: this.compile,
          renderToString: this.renderToString,
        }}
      >
        {children}
      </Provider>
    );
  }
}

function Message(props: MessageProps) {
  try {
    const { id, values, components, language, compile } = props;
    const tokens = compile(id);
    const result = evaluate(tokens, language, values || {}, components || {});
    const children = [];
    for (let i = 0; i < result.length; ++i) {
      const element = result[i];
      children.push(<React.Fragment key={i}>{element}</React.Fragment>);
    }
    return <>{children}</>;
  } catch (e) {
    console.warn(e);
    return ("" as any) as React.ReactElement<any>;
  }
}

export class FormattedMessage extends React.Component<MessageOwnProps> {
  renderMessage = (contextValue: ContextValue) => {
    return <Message {...this.props} {...contextValue} />;
  };

  render() {
    return <Consumer>{this.renderMessage}</Consumer>;
  }
}

export function useMessageFormat(props: Omit<MessageOwnProps, "id"> = {}) {
  const { useState, useContext, useCallback } = React;

  if (!useState) {
    throw new Error("At least react@16.7 is required to use useMessageFormat");
  }

  const context = useContext(Context);

  const compile = useCallback(
    (
      id: MessageOwnProps["id"],
      values?: MessageOwnProps["values"],
      components?: MessageOwnProps["components"]
    ) => (
      <FormattedMessage
        {...props}
        id={id}
        values={values || props.values}
        components={components || props.components}
      />
    ),
    [props, context]
  );

  return {
    ...context,
    compile,
  };
}
