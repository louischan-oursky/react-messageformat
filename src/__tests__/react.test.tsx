import * as React from "react";
import * as renderer from "react-test-renderer";
import {
  LocaleProvider,
  Context,
  FormattedMessage,
  MessageOwnProps,
  useMessageFormat,
} from "../react";

const locale = "en";
const messageByID = {
  "plain.string": "Hello world",
  argument: "I meet {GUEST}",
  plural: "{N, plural, =4{four} one{ichi} other{#}}",
  "react.element": "{ANY} {NESTED} {REACT} {ELEMENT}",
  react: "This is a {A, react, href{{SCHEME}://{HOST}} children{link}}",
  "react.nonstring": "This is a {A, react, href{{href}}}",
};

function Shortcut(props: MessageOwnProps) {
  return (
    <LocaleProvider locale={locale} messageByID={messageByID}>
      <FormattedMessage {...props} />
    </LocaleProvider>
  );
}

function Input(props: MessageOwnProps) {
  return (
    <Context.Consumer>
      {({ renderToString }) => {
        return <input placeholder={renderToString(props.id, props.values)} />;
      }}
    </Context.Consumer>
  );
}

function UseMessageFormatPlainString(props: MessageOwnProps) {
  const messageFormat = useMessageFormat();

  return <input value={messageFormat.renderToString(props.id, props.values)} />;
}

function UseMessageFormatEmbedded(props: MessageOwnProps) {
  const messageFormat = useMessageFormat();

  return messageFormat.compile(props.id, props.values, props.components);
}

class AnyComponent extends React.Component {
  render() {
    return "AnyComponent";
  }
}

test("plain.string", () => {
  const tree = renderer.create(<Shortcut id="plain.string" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("argument", () => {
  const tree = renderer
    .create(
      <Shortcut
        id="argument"
        values={{
          GUEST: "John",
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("plural", () => {
  const tree = renderer
    .create(
      <Shortcut
        id="plural"
        values={{
          N: 4,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("React Element as value", () => {
  const tree = renderer
    .create(
      <LocaleProvider locale={locale} messageByID={messageByID}>
        <FormattedMessage
          id="react.element"
          values={{
            ANY: <FormattedMessage id="plain.string" />,
            NESTED: <AnyComponent />,
            REACT: (
              <FormattedMessage
                id="argument"
                values={{
                  GUEST: "John",
                }}
              />
            ),
            ELEMENT: (
              <FormattedMessage
                id="plural"
                values={{
                  N: 1,
                }}
              />
            ),
          }}
        />
      </LocaleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Embedded React Element", () => {
  const tree = renderer
    .create(
      <LocaleProvider locale={locale} messageByID={messageByID}>
        <FormattedMessage
          id="react"
          components={{
            A: "a",
          }}
          values={{
            SCHEME: "https",
            HOST: "www.example.com",
          }}
        />
      </LocaleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Embedded React Element with non-string value", () => {
  const tree = renderer
    .create(
      <LocaleProvider locale={locale} messageByID={messageByID}>
        <FormattedMessage
          id="react.nonstring"
          components={{
            A: "a",
          }}
          values={{
            href: {
              valueOf: () => "https://reactjs.org",
              toString: () => "https://reactjs.org",
            },
          }}
        />
      </LocaleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("imperative", () => {
  const tree = renderer
    .create(
      <LocaleProvider locale={locale} messageByID={messageByID}>
        <Input id="plain.string" />
      </LocaleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Hook with plain string", () => {
  const tree = renderer
    .create(
      <LocaleProvider locale={locale} messageByID={messageByID}>
        <UseMessageFormatPlainString id="plain.string" />
      </LocaleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Hook with React Element", () => {
  const tree = renderer
    .create(
      <LocaleProvider locale={locale} messageByID={messageByID}>
        <UseMessageFormatEmbedded
          id="react.element"
          values={{
            ANY: <FormattedMessage id="plain.string" />,
            NESTED: <AnyComponent />,
            REACT: (
              <FormattedMessage
                id="argument"
                values={{
                  GUEST: "John",
                }}
              />
            ),
            ELEMENT: (
              <FormattedMessage
                id="plural"
                values={{
                  N: 1,
                }}
              />
            ),
          }}
        />
        <UseMessageFormatEmbedded
          id="react"
          components={{
            A: "a",
          }}
          values={{
            SCHEME: "https",
            HOST: "www.example.com",
          }}
        />
      </LocaleProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
