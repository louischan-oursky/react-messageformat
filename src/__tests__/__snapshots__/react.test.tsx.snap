// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Embedded React Element 1`] = `
Array [
  "This is a ",
  <a
    href="https://www.example.com"
  >
    link
  </a>,
]
`;

exports[`Embedded React Element with non-string value 1`] = `
Array [
  "This is a ",
  <a
    href={
      Object {
        "toString": [Function],
        "valueOf": [Function],
      }
    }
  />,
]
`;

exports[`Hook with React Element 1`] = `
Array [
  "Hello world",
  " ",
  "AnyComponent",
  " ",
  "I meet ",
  "John",
  " ",
  "ichi",
  "This is a ",
  <a
    href="https://www.example.com"
  >
    link
  </a>,
]
`;

exports[`Hook with plain string 1`] = `
<input
  value="Hello world"
/>
`;

exports[`React Element as value 1`] = `
Array [
  "Hello world",
  " ",
  "AnyComponent",
  " ",
  "I meet ",
  "John",
  " ",
  "ichi",
]
`;

exports[`argument 1`] = `
Array [
  "I meet ",
  "John",
]
`;

exports[`imperative 1`] = `
<input
  placeholder="Hello world"
/>
`;

exports[`plain.string 1`] = `"Hello world"`;

exports[`plural 1`] = `"four"`;
